libgadget/treewalk.c

```diff
@@ -560,25 +522,35 @@ int treewalk_export_particle(LocalTreeWalk * lv, int no) {
 
     if(exportnodecount[task] == NODELISTLENGTH)
     {
-        int nexp;
-
-        if(tw->Nexport < tw->BunchSize) {
-            nexp = atomic_fetch_and_add(&tw->Nexport, 1);
-        } else {
-            nexp = tw->BunchSize;
-        }
+        const int nexp = atomic_fetch_and_add(&tw->Nexport, 1);
 
+        /* out of buffer space. Need to discard work for this particle and interrupt */
         if(nexp >= tw->BunchSize) {
-            /* out of buffer space. Need to discard work for this particle and interrupt */
             tw->BufferFullFlag = 1;
+            /* This reduces the time until the other threads see the buffer is full and the loop can exit.
+             * Since it is a pure optimization, no need for a full atomic.*/
 #pragma omp flush
+            /* Touch up the DataIndexTable, so that exports associated with the current particle
+             * won't be exported. This is expensive but rare. */
+            int i;
+            for(i=0; i < tw->BunchSize; i++) {
```

Comment:
Looks like you are disgarding the entire bunch instead of the last particle?

Reply:
I don't think so - isn't lv->target set to the current particle? So I am discarding only elements of the Bunch which are related to the current particle.

Reply:
If you confirm that I haven't misunderstood the code, I will add a comment.

Reply:
Yes. I agree with you on this. target is the only item that we have failed to export from this thread, thus this search is sufficient to revert the partial export of target.

Reply:
Thanks, in that case I will merge this.